lib/rust/ensogl/component/flame-graph/src/mark.rs

```diff
@@ -0,0 +1,190 @@
+            let width  : Var<Pixels> = MARK_WIDTH.px();
+            let height : Var<Pixels> = INFINITE.px();
+            let zoom                 = &Var::<f32>::from("1.0/zoom()");
+            let base_color           = style.get_color("flame_graph_mark_color");
+
+            let shape = Rect((&width*zoom,&height));
+            let shape = shape.fill(base_color);
+
+            let hover_area = shape.grow(zoom * MARK_HOVER_AREA_WIDTH.px());
+            let hover_area = hover_area.fill(HOVER_COLOR);
+
+
+            (shape + hover_area).into()
+        }
+    }
+}
+
+
+
+// ===========
+// === FRP ===
+// ===========
+
+ensogl_core::define_endpoints_2! {
+    Input {
+        set_content(String),
+    }
+    Output {}
+}
+
+impl component::Frp<Model> for Frp {
+    fn init(api: &Self::Private, app: &Application, model: &Model, _style: &StyleWatchFrp) {
+        let network = &api.network;
+        let background = &model.background.events;
+
+        let cursor_pos = app.cursor.frp.scene_position.clone_ref();
+        frp::extend! { network
+            eval api.input.set_content((t) model.set_content(t));
+
+            is_hovered <- bool(&background.mouse_out, &background.mouse_over);
+            eval is_hovered((hovered) model.set_label_visible(*hovered));
+
+            on_mouse_over_pos <- cursor_pos.gate(&is_hovered);
+            eval on_mouse_over_pos((pos) model.set_label_y(pos.y));
+        }
+
+        model.set_size(Vector2::new(MARK_WIDTH + 2.0 * MARK_HOVER_AREA_WIDTH, INFINITE));
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+#[derive(Clone, CloneRef, Debug)]
+pub struct Model {
+    app:            Application,
+    background:     background::View,
+    label:          Rc<RefCell<Option<text::Area>>>,
+    display_object: display::object::Instance,
+    text:           Rc<RefCell<Option<String>>>,
+}
+
+impl component::Model for Model {
+    fn label() -> &'static str {
+        "FlameGraphMark"
+    }
+
+    fn new(app: &Application, logger: &Logger) -> Self {
+        let scene = &app.display.default_scene;
+        let display_object = display::object::Instance::new(&logger);
+        let label = default();
+        let text = default();
+
+        let background = background::View::new(&logger);
+        display_object.add_child(&background);
+        scene.layers.tooltip.add_exclusive(&background);
+
+        let app = app.clone_ref();
+        Model { app, background, label, display_object, text }
+    }
+}
+
+impl Model {
+    fn set_size(&self, size: Vector2) {
+        self.background.size.set(size);
+    }
+
+    fn set_content(&self, t: &str) {
+        self.text.set(t.to_owned());
+        if let Some(label) = self.label.borrow().deref() {
+            label.set_content(t.to_owned())
+        }
+    }
+
+    fn set_label_visible(&self, visible: bool) {
+        if visible {
+            self.enable_label();
+        } else if let Some(label) = self.label.take() {
```

Comment:
The condition test does the main work here. I think it would be more clear for an unconditional else to do `.take().for_each(...)`